base_user_role_history/models/res_users.py

```diff
@@ -0,0 +1,75 @@
+# -*- coding: utf-8 -*-
+# Copyright 2019 ACSONE SA/NV
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models, _
+
+
+class ResUsers(models.Model):
+    _inherit = 'res.users'
+
+    last_role_line_modification = fields.Datetime(
+        string="Last roles modification",
+        readonly=True,
+    )
+
+    @api.model
+    def _prepare_role_line_history_dict(self, role_line):
+        return {
+            'user_id': role_line.user_id.id,
+            'role_id': role_line.role_id.id,
+            'date_from': role_line.date_from,
+            'date_to': role_line.date_to,
+            'is_enabled': role_line.is_enabled,
+        }
+
+    @api.multi
+    def _get_role_line_values_by_user(self):
+        role_line_values_by_user = {}
+        for rec in self:
+            role_line_values_by_user.setdefault(rec, {})
+            for role_line in rec.role_line_ids:
+                role_line_values_by_user[rec][role_line.id] = \
+                    self._prepare_role_line_history_dict(role_line)
+        return role_line_values_by_user
+
+    @api.model
```

Comment:
Better using `@api.model_create_multi` starting from Odoo >= 12.0 to receive a list of values, and rename `vals` to `vals_list` (https://github.com/odoo/odoo/blob/12.0/odoo/models.py#L3471)